### PR TITLE
Feat: Improve hub integration

### PR DESCRIPTION
**Pull request by pkooij**
## What this does
This PR adds 
- integration to automatically push the policy to the Hub after training
- Adds the correct tags for the policies
- Add a model card for a policy.
- If applicable add dataset used to train and the base_model (in case of smolvla)

**Example policy upload -> [link](https://huggingface.co/pepijn223/my_policy50)**

### To test
Run
```bash
python lerobot/scripts/train.py \
  --policy.type=smolvla \
  --dataset.repo_id=imstevenpmwork/thanos_picking_power_gem_1749731584242992 \
  --batch_size=16 \
  --steps=1 \
  --policy.device=mps \
  --policy.repo_id=${HF_USER}/my_policy21 \
  --wandb.enable=true
```

cc @Vaibhavs10

### CODE REVIEW

File: docs/source/il_robots.mdx [hunk 1]
```diff
@@ -273,6 +274,8 @@ python lerobot/scripts/train.py \
   --resume=true
 ```
 
+If you do not want to push your model to the hub after training use `--policy.push_to_hub=false`.
```
Comment (Vaibhavs10) [inline docs/source/il_robots.mdx:277]: nice to make it the default

File: lerobot/configs/train.py [hunk 2]
```diff
@@ -128,6 +133,33 @@ def _save_pretrained(self, save_directory: Path) -> None:
...
+    def generate_model_card(self) -> ModelCard:
+        dataset_repo_id = self.dataset.repo_id
+        datasets = [dataset_repo_id] if dataset_repo_id and isinstance(dataset_repo_id, str) else None
+
+        model_type = self.policy.type  # This is the policy name
+        base_model = "lerobot/smolvla_base" if model_type == "smolvla" else None  # Set a base model
+
+        card_data = ModelCardData(
+            license="apache-2.0",
+            library_name="lerobot",
```
Comment (Vaibhavs10) [inline lerobot/configs/train.py]: nice!

File: lerobot/templates/lerobot_modelcard_template.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+{% else %}
+_Model type not recognized — please update this template._
+{% endif %}
+
+This policy has been trained and pushed to the Hub using [LeRobot](https://github.com/huggingface/lerobot).
+See the full documentation at [LeRobot Docs](https://huggingface.co/docs/lerobot/index).
+
+---
+
+## How to Get Started with the Model
```
Comment (Vaibhavs10) [inline lerobot/templates/lerobot_modelcard_template.md:38]: Perhaps we could add a more nicer inference snippet here

File: lerobot/common/utils/train_utils.py [hunk 4]
```diff
@@ -159,3 +159,24 @@ def load_training_state(
         scheduler = load_scheduler_state(scheduler, training_state_dir)
 
     return step, optimizer, scheduler
+
+
+def push_policy_to_hub(cfg: TrainPipelineConfig, policy: PreTrainedPolicy):
+    url = policy.push_to_hub(
+        repo_id=cfg.policy.repo_id,
+        private=False,
```
Comment (Wauplin) [inline lerobot/common/utils/train_utils.py]:
```suggestion
```

Better to keep the default value (which is `None`). In practice it defaults to `private=False` implicitly for most users but one of the Enterprise Hub feature is to create all repos as private by default. If we pass `private=False` here, the Enterprise Hub feature won't work

File: lerobot/common/utils/train_utils.py [hunk 5]
```diff
@@ -159,3 +159,24 @@ def load_training_state(
...
+        branch="main",
+        allow_patterns=["*.safetensors", "*.json", "*.yaml", "*.md"],
+        ignore_patterns=["*.tmp", "*.log"],
+    )
+
+    cfg.push_to_hub(
+        repo_id=cfg.policy.repo_id,
+        private=False,
```
Comment (Wauplin) [inline lerobot/common/utils/train_utils.py]:
```suggestion
```

same here

File: lerobot/common/utils/train_utils.py [hunk 6]
```diff
@@ -159,3 +159,24 @@ def load_training_state(
         scheduler = load_scheduler_state(scheduler, training_state_dir)
 
     return step, optimizer, scheduler
+
+
+def push_policy_to_hub(cfg: TrainPipelineConfig, policy: PreTrainedPolicy):
+    url = policy.push_to_hub(
+        repo_id=cfg.policy.repo_id,
+        private=False,
+        branch="main",
```
Comment (Wauplin) [inline lerobot/common/utils/train_utils.py]:
```suggestion
```

(not needed and therefore better to default to None)

File: lerobot/common/utils/train_utils.py [hunk 7]
```diff
@@ -159,3 +159,24 @@ def load_training_state(
...
+        allow_patterns=["*.safetensors", "*.json", "*.yaml", "*.md"],
+        ignore_patterns=["*.tmp", "*.log"],
+    )
+
+    cfg.push_to_hub(
+        repo_id=cfg.policy.repo_id,
+        private=False,
+        branch="main",
```
Comment (Wauplin) [inline lerobot/common/utils/train_utils.py]:
```suggestion
```

(not needed)

File: lerobot/configs/train.py (hunk 2 above)
Comment (Wauplin) [reply lerobot/configs/train.py]: :tada:

File: lerobot/templates/lerobot_modelcard_template.md
Comment (Wauplin) [inline lerobot/templates/lerobot_modelcard_template.md:1]: This is not a `.py` file, meaning that if someone install `lerobot` from PyPI or build the package themselves, the file won't be present. To tell poetry to add it when building the package, you should add a `MANIFEST.in` file like this: 

```
# MANIFEST.in
include lerobot/templates/lerobot_modelcard_template.md"
``` 

Here is the example in `huggingface_hub`: https://github.com/huggingface/huggingface_hub/blob/main/MANIFEST.in.

(note: I did not test it myself but it seems you can tweak your `pyproject.toml` file to do the same https://stackoverflow.com/a/76485475)

Comment (pkooij) [reply lerobot/templates/lerobot_modelcard_template.md:1]: Thanks! @imstevenpmwork do you want me to already add a MANIFEST.in or do you want to take it into account when adding PyPI build

File: lerobot/templates/lerobot_modelcard_template.md (hunk 3 above)
Comment (pkooij) [reply lerobot/templates/lerobot_modelcard_template.md:38]: I will add this when we also include a API example for inference and eval

File: lerobot/common/utils/train_utils.py (hunk 4 above)
Comment (pkooij) [reply lerobot/common/utils/train_utils.py]: Thanks good catch, removed them in commit 825c1dac849ecb3124280d5f5578996d171077e7

File: lerobot/common/utils/train_utils.py [hunk 8]
```diff
@@ -159,3 +159,20 @@ def load_training_state(
         scheduler = load_scheduler_state(scheduler, training_state_dir)
 
     return step, optimizer, scheduler
+
+
+def push_policy_to_hub(cfg: TrainPipelineConfig, policy: PreTrainedPolicy):
```
Comment (imstevenpmwork) [inline lerobot/common/utils/train_utils.py]: Consider instead creating a method in the policy class which does all the pushing related logic: creates the model card, takes the TrainPipelineConfig object and invokes its `push_to_hub` method, and calls its own `self.push_to_hub`.
I prefer this over having an util function because it is closer from what we have for the dataset upload.

In addition to this, users should be able to choose tags, license and so on the same way they can for the datasets.

File: lerobot/configs/policies.py [hunk 9]
```diff
@@ -60,6 +60,10 @@ class PreTrainedConfig(draccus.ChoiceRegistry, HubMixin, abc.ABC):
     # automatic gradient scaling is used.
     use_amp: bool = False
 
+    push_to_hub: bool = True
+    repo_id: str | None = None
+    dataset_repo_id: str | None = None
```
Comment (imstevenpmwork) [inline lerobot/configs/policies.py]: I believe `dataset_repo_id` is unused

File: lerobot/configs/train.py [hunk 10]
```diff
@@ -128,6 +133,33 @@ def _save_pretrained(self, save_directory: Path) -> None:
         with open(save_directory / TRAIN_CONFIG_NAME, "w") as f, draccus.config_type("json"):
             draccus.dump(self, f, indent=4)
 
+        card = self.generate_model_card()
+        card.save(str(save_directory / "README.md"))
+
+        self.policy.save_pretrained(save_directory=save_directory)  # Also push model
+
+    def generate_model_card(self) -> ModelCard:
```
Comment (imstevenpmwork) [inline lerobot/configs/train.py]: I have a slight preference to have this generation function in the policy class instead. Just because this is a config class, and its scope should be limited to how we handle options for training.

File: lerobot/scripts/train.py [hunk 11]
```diff
@@ -282,6 +283,9 @@ def train(cfg: TrainPipelineConfig):
         eval_env.close()
     logging.info("End of training")
 
+    if cfg.policy.push_to_hub and cfg.policy.repo_id:
```
Comment (imstevenpmwork) [inline lerobot/scripts/train.py]: In my opinion, the check for `cfg.policy.repo_id` should happen in the function to be called. Ideally I picture something like this:

```python
if policy.push_to_hub:
    policy.push_model_to_hub(cfg, ...) # cfg being TrainPipelineConfig which we need for uploading train_config.json
```

File: lerobot/templates/lerobot_modelcard_template.md
Comment (imstevenpmwork) [reply lerobot/templates/lerobot_modelcard_template.md:1]: This is a good catch, thanks @Wauplin !
@pkooij If you can add it then it would be great, just consider that the manifest would also need to include the card template we use for the datasets

**Review by imstevenpmwork — COMMENTED** [review]
We discussed already in person but left some comments for visibility. I think this will be an impactful addition for helping us keep track of the adoption trends, thanks!

File: lerobot/configs/train.py (hunk 10 above)
Comment (pkooij) [reply lerobot/configs/train.py]: I checked it, we need the dataset repo id which is not available on the policy that's why this generate is in train

File: lerobot/common/utils/train_utils.py (hunk 8 above)
Comment (pkooij) [reply lerobot/common/utils/train_utils.py]: Added this to PreTrainedPolicy in train.py you can now also do: `--policy.private=true --policy.tags=\[ppo,rl\] --policy.license=mit` for example

File: lerobot/configs/policies.py (hunk 9 above)
Comment (pkooij) [reply lerobot/configs/policies.py]: Correct, removed it

File: lerobot/scripts/train.py (hunk 11 above)
Comment (pkooij) [reply lerobot/scripts/train.py]: Moved it

File: lerobot/templates/lerobot_modelcard_template.md
Comment (pkooij) [reply lerobot/templates/lerobot_modelcard_template.md:1]: Added

File: lerobot/configs/policies.py [hunk 12]
```diff
@@ -60,6 +60,16 @@ class PreTrainedConfig(draccus.ChoiceRegistry, HubMixin, abc.ABC):
     # automatic gradient scaling is used.
     use_amp: bool = False
 
+    push_to_hub: bool = True
+    repo_id: str | None = None
+
+    # Upload on private repository on the Hugging Face hub.
+    private: bool = False
```
Comment (imstevenpmwork) [inline lerobot/configs/policies.py]: I know it seems counterintuitive, but base on this [comment](https://github.com/huggingface/lerobot/pull/1382#discussion_r2166642425) I believe we should do this instead:
```suggestion
    private: bool | None = None
```

File: lerobot/common/policies/pretrained.py [hunk 13]
```diff
@@ -197,3 +179,60 @@ def select_action(self, batch: dict[str, Tensor]) -> Tensor:
...
         """
         raise NotImplementedError
+
+    def push_model_to_hub(
+        self,
+        cfg: TrainPipelineConfig,
+    ):
+        if not cfg.policy.repo_id:
+            logging.warning("`policy.repo_id` is not specified, please specify it to use push_model_to_hub()")
+            return
```
Comment (imstevenpmwork) [inline lerobot/common/policies/pretrained.py]: To keep it consistent with other parts of the code, I would raise here instead:
```suggestion
        if not self.config.repo_id:
            raise ValueError("'policy.repo_id' argument missing. Please specify it to push the model to the hub.")
```
Also, prioritize the use of `self` as you already have the config in here

File: lerobot/common/policies/pretrained.py [hunk 14]
```diff
@@ -197,3 +179,60 @@ def select_action(self, batch: dict[str, Tensor]) -> Tensor:
...
+
+        api = HfApi()
+        repo_id = api.create_repo(
+            repo_id=cfg.policy.repo_id, private=cfg.policy.private, exist_ok=True
+        ).repo_id
```
Comment (imstevenpmwork) [inline lerobot/common/policies/pretrained.py]:
```suggestion
        repo_id = api.create_repo(
            repo_id=self.config.repo_id, private=self.config.private, exist_ok=True
        ).repo_id
```

File: lerobot/common/policies/pretrained.py [hunk 15]
```diff
@@ -197,3 +179,60 @@ def select_action(self, batch: dict[str, Tensor]) -> Tensor:
...
+
+        # Push the files to the repo in a single commit
+        with TemporaryDirectory(ignore_cleanup_errors=True) as tmp:
+            saved_path = Path(tmp) / repo_id
+
+            self.save_pretrained(saved_path)  # Calls _save_pretrained and stores model tensors
+
+            card = self.generate_model_card(
+                cfg.dataset.repo_id, cfg.policy.type, cfg.policy.license, cfg.policy.tags
```
Comment (imstevenpmwork) [inline lerobot/common/policies/pretrained.py]:
```suggestion
                cfg.dataset.repo_id, self.config.type, self.config.license, self.config.tags
```

File: lerobot/common/policies/pretrained.py [hunk 16]
```diff
@@ -197,3 +179,60 @@ def select_action(self, batch: dict[str, Tensor]) -> Tensor:
...
+    def generate_model_card(
+        self, dataset_repo_id: str, model_type: str, license: str | None, tags: List[str] | None
+    ) -> ModelCard:
+        base_model = "lerobot/smolvla_base" if model_type == "smolvla" else None  # Set a base model
+
+        card_data = ModelCardData(
+            license=license or "apache-2.0",
+            library_name="lerobot",
+            pipeline_tag="robotics",
+            tags=tags or ["robotics", model_type],
```
Comment (imstevenpmwork) [inline lerobot/common/policies/pretrained.py]:
```suggestion
            tags=list(set(tags or []).union({"robotics", "lerobot", model_type})),
```

File: lerobot/common/policies/pretrained.py [hunk 17]
```diff
@@ -197,3 +179,60 @@ def select_action(self, batch: dict[str, Tensor]) -> Tensor:
...
+            model_name=model_type,
+            datasets=dataset_repo_id,
+            base_model=base_model,
+        )
+
+        template_path = files("lerobot.templates").joinpath("lerobot_modelcard_template.md")
+        card = ModelCard.from_template(card_data, template_path=str(template_path))
```
Comment (imstevenpmwork) [inline lerobot/common/policies/pretrained.py]:
```suggestion
        template_card = files("lerobot.templates").joinpath("lerobot_modelcard_template.md").read_text()
        card = ModelCard.from_template(card_data, template_str=template_card)
```

File: lerobot/common/policies/pretrained.py [hunk 18]
```diff
@@ -197,3 +179,60 @@ def select_action(self, batch: dict[str, Tensor]) -> Tensor:
...
+            url = api.upload_folder(
+                repo_id=repo_id,
+                repo_type="model",
+                folder_path=saved_path,
+                commit_message="Upload policy weights, train config and readme",
+                allow_patterns=["*.safetensors", "*.json", "*.yaml", "*.md"],
+                ignore_patterns=["*.tmp", "*.log"],
+            )
+
+            logging.info(f"Model pushed to {url}")
```
Comment (imstevenpmwork) [inline lerobot/common/policies/pretrained.py]: Just checking, are you sure this is doing what you expect? The return type of `upload_folder` is a `CommitInfo` struct.

File: lerobot/common/policies/pretrained.py (hunk 13 above)
Comment (imstevenpmwork) [reply lerobot/common/policies/pretrained.py]: I just noticed that we have a `cfg.validate()`, I think it would make more sense to do this check over there. Why? We fail fast in case we want to push but the repo_id is not specified. Otherwise we will raise at the very end and the user will have to upload the model as we currently do so (via the `huggingface-cli` tool.

File: lerobot/common/policies/pretrained.py (hunk 18 above)
Comment (pkooij) [reply lerobot/common/policies/pretrained.py]: Now only print url, good catch

**Review by imstevenpmwork — DISMISSED** [review]
LGTM
The test are failing because we need to set `push_to_hub` to `False` in the tests. I suggest overwriting the default in every instance of `make_policy_config` in the test folder e.g. `policy_cfg = make_policy_config(policy_name, push_to_hub=False)` should do the trick!
